### PR TITLE
Auto: implement issue #510

### LINKED ISSUE #510: Add customer_enrichments ORM model and Alembic migration
Subtask of #74

Add `src/db/models/customer_enrichment.py` with the `CustomerEnrichment` model (id, customer_id FK, provider, raw_data_json, enriched_at, next_refresh_at) and import it in `alembic/env.py`. Create an alembic migration for the new table with proper indexes on customer_id and next_refresh_at. Write a unit test that imports the model and verifies the table can be created. No external API calls.

### PR DESCRIPTION
**Pull request by yihua-zhuo**
Closes #510

Plan: `.plans/issue-510.md`

This PR was auto-implemented by Claude Code in three stages:

1. **Plan** — Claude wrote a detailed implementation plan to `.plans/issue-510.md`.
2. **Implement** — Claude executed the plan, ran `ruff check src/` + the unit and integration test suites, and iterated until all three were green.
3. **Self-review** — Claude reviewed its own diff against `codereview.md` (the project review checklist) and fixed any rule violations found, re-running the test suite if any changes were made.

PR-time CI (`Dev Agent CI`) re-verifies independently before this PR is mergeable.

Please review before merging.

## Summary by CodeRabbit

* **New Features**
  * Added customer enrichment data storage system enabling persistent storage of enriched customer records with flexible metadata fields, enrichment tracking timestamps, and automatic refresh scheduling capabilities.

* **Documentation**
  * Added comprehensive implementation plan documentation for the customer enrichment feature including requirements, acceptance criteria, and technical design specifications.

[Review Change Stack](https://app.coderabbit.ai/change-stack/yihua-zhuo/agent-job/pull/656?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: .plans/issue-510.md [hunk 1]
```diff
@@ -0,0 +1,30 @@
+Now I have everything needed. Let me write the plan.
+
+# Implementation Plan — Issue #510
+
+## Goal
+Add a `CustomerEnrichment` ORM model and corresponding Alembic migration to persist third-party customer data enrichment records. The model stores enrichment provider, raw JSON payload, and refresh scheduling, linked to a customer via a foreign key with indexes for efficient querying.
+
+## Affected Files
+- `src/db/models/customer_enrichment.py` — New file: `CustomerEnrichment` ORM model (id, customer_id FK, provider, raw_data_json, enriched_at, next_refresh_at)
```
Comment (yihua-zhuo) [inline .plans/issue-510.md:9]: **CRITICAL** (critical)

**Issue:** Plan lists affected files but omits a `tenant_id` column for the `customer_enrichments` table, contradicting Rule 126 (mandatory tenant isolation) and Rule 138 (tenant-owned ORM models need a proper foreign key to `tenants.id`). Every other tenant-owned table in this codebase (e.g. `opportunities`, `automation_rules`, `automation_logs`) has an explicit `tenant_id` column.

**Suggestion:** Add `tenant_id: Mapped[int] = mapped_column(Integer, default=0, nullable=False, index=True)` to the model and `sa.Column('tenant_id', sa.Integer(), default=0, nullable=False)` to the migration, with a foreign key to `tenants.id`.

File: tests/unit/test_customer_enrichment_model.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+        """Test enrichment creation with required fields and default values."""
+        enrichment = CustomerEnrichmentModel(
+            customer_id=1,
+            provider="clearbit",
+        )
+
+        assert enrichment.customer_id == 1
+        assert enrichment.provider == "clearbit"
+        # Python-level default is not applied on construction; to_dict() handles it
+        assert enrichment.raw_data_json is None
```
Comment (yihua-zhuo) [inline tests/unit/test_customer_enrichment_model.py:41]: **CRITICAL** (critical)

**Issue:** Test asserts `enrichment.raw_data_json is None` (line 41) and `enrichment.raw_data_json is None` (line 89). The model has `default=dict, nullable=False` on the `raw_data_json` column; SQLAlchemy's `__init__` does not apply Python column defaults on bare construction, so the attribute would raise `AttributeError` or behave inconsistently rather than be `None`. The test's own comment admits 'Python-level default is not applied on construction', meaning the assertions check undefined/misleading state rather than documented behavior.

**Suggestion:** Assert the documented Python-level default: `assert enrichment.raw_data_json == {}` after bare construction. Add a separate test that `to_dict()` converts `None` → `{}` (already covered in `test_raw_data_json_default` lines 89-90 — fix the assertion there too).

File: .plans/issue-510.md [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+- `alembic/versions/<revision>_create_customer_enrichments.py` — New file: Alembic migration creating the `customer_enrichments` table with indexes on `customer_id` and `next_refresh_at`
+- `tests/unit/test_customer_enrichment_model.py` — New file: unit tests verifying model instantiation and `to_dict()` serialization
+
+## Implementation Steps
+1. **Create `src/db/models/customer_enrichment.py`**: Define `CustomerEnrichmentModel` inheriting from `db.base.Base`, mirroring the pattern in `pipeline_stage.py`. Fields: `id` (Integer PK, autoincrement), `customer_id` (Integer, ForeignKey `customers.id`, nullable=False, index=True), `provider` (String(100), nullable=False), `raw_data_json` (JSON, default=dict, nullable=False), `enriched_at` (DateTime timezone=True, nullable=True), `next_refresh_at` (DateTime timezone=True, nullable=True, index=True). Add standard `created_at`/`updated_at` with `server_default=func.now()`. Implement `to_dict()` serializing all fields, with ISO format for datetimes.
```
Comment (yihua-zhuo) [inline .plans/issue-510.md:14]: **WARNING** (suggestion)

**Issue:** Step 1 says "JSON, default=dict" but the implementation uses `JSONB`. Step 2 also says "JSON vs JSONB — if enrichment queries ever need to index inside the JSON payload, JSONB is preferred", implying the decision was left open. The plan should record which type was chosen and why.

**Suggestion:** Update Step 1 to read "raw_data_json (postgresql.JSONB, default=dict)" and resolve the open risk by documenting the chosen type (JSONB) in the Risks section.

File: .plans/issue-510.md [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+2. **Create the Alembic migration**: Run `alembic revision --autogenerate -m "create_customer_enrichments"` targeting the `customers` table (revising from `c94d682d4b03`). Manually add an index on `next_refresh_at` (`ix_customer_enrichments_next_refresh_at`) alongside the autogenerated `customer_id` index. Delete any empty follow-up "drift_check" migration.
+3. **Create `tests/unit/test_customer_enrichment_model.py`**: Add a `TestCustomerEnrichmentModel` class with `test_create_enrichment_defaults`, `test_to_dict`, and `test_customer_id_foreign_key` test methods following the style in `test_opportunity_model.py`.
```
Comment (yihua-zhuo) [inline .plans/issue-510.md:16]: **INFO** (suggestion)

**Issue:** Step 2 references `c94d682d4b03` as the autogenerate base — this is correct (confirmed in the migration file's `down_revision`), but the plan text does not call out that autogenerate was run manually against a disposable DB rather than the test DB, which is a key gotcha from CLAUDE.md.

**Suggestion:** Add a note that autogenerate was run against a dedicated disposable `alembic_dev` database, not `test_db`, to avoid phantom diffs.

File: .plans/issue-510.md [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+- Unit tests in `tests/unit/`: `test_customer_enrichment_model.py` — instantiates `CustomerEnrichmentModel` directly (no DB), asserts fields, enum-style values, and `to_dict()` output.
+- Integration tests in `tests/integration/`: None required — the issue explicitly disclaims external API calls and limits scope to unit-level table-creation verification.
+
+## Acceptance Criteria
+- `src/db/models/customer_enrichment.py` imports cleanly with `python -c "from src.db.models import CustomerEnrichmentModel"` (PYTHONPATH=src).
+- `alembic/versions/` contains a migration that `alembic upgrade head` applies without error against a live PostgreSQL DB.
+- `alembic downgrade -1` reverses the migration cleanly.
+- Unit test `pytest tests/unit/test_customer_enrichment_model.py -v` passes with zero errors.
+- `ruff check src/db/models/customer_enrichment.py` and `mypy src/db/models/customer_enrichment.py` are clean.
+
```
Comment (yihua-zhuo) [inline .plans/issue-510.md:28]: **INFO** (suggestion)

**Issue:** Acceptance criterion says "ruff check src/db/models/customer_enrichment.py" but Rule 145 requires fenced blocks in markdown to have explicit language tags. The criterion is plain text without a `bash` fence, which is fine as inline command reference but is inconsistent with the project's own markdown conventions.

**Suggestion:** Wrap the ruff/mypy commands in fenced bash blocks: ```bash\nruff check src/db/models/customer_enrichment.py\n```

File: alembic/versions/f18b406b982a_create_customer_enrichments.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+        sa.Column('customer_id', sa.Integer(), nullable=False),
+        sa.Column('provider', sa.String(length=100), nullable=False),
+        sa.Column('raw_data_json', postgresql.JSONB(astext_type=sa.Text()), nullable=False),
+        sa.Column('enriched_at', sa.DateTime(timezone=True), nullable=True),
+        sa.Column('next_refresh_at', sa.DateTime(timezone=True), nullable=True),
+        sa.Column('created_at', sa.DateTime(timezone=True), server_default=sa.text('now()'), nullable=False),
+        sa.Column('updated_at', sa.DateTime(timezone=True), server_default=sa.text('now()'), nullable=False),
+        sa.ForeignKeyConstraint(['customer_id'], ['customers.id'], ondelete='CASCADE'),
+        sa.PrimaryKeyConstraint('id'),
+        sqlite_autoincrement=True,
```
Comment (yihua-zhuo) [inline alembic/versions/f18b406b982a_create_customer_enrichments.py:35]: **INFO** (suggestion)

**Issue:** `sqlite_autoincrement=True` is a SQLite-specific directive; this is a PostgreSQL migration. It has no functional effect on PostgreSQL but indicates the migration was written with a cross-database template.

**Suggestion:** Remove `sqlite_autoincrement=True` from `create_table()` since it is unnecessary for PostgreSQL.

File: alembic/versions/f18b406b982a_create_customer_enrichments.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+depends_on: str | Sequence[str] | None = None
+
+
+def upgrade() -> None:
+    op.create_table(
+        'customer_enrichments',
+        sa.Column('id', sa.Integer(), autoincrement=True, nullable=False),
+        sa.Column('customer_id', sa.Integer(), nullable=False),
+        sa.Column('provider', sa.String(length=100), nullable=False),
+        sa.Column('raw_data_json', postgresql.JSONB(astext_type=sa.Text()), nullable=False),
```
Comment (yihua-zhuo) [inline alembic/versions/f18b406b982a_create_customer_enrichments.py:28]: **WARNING** (suggestion)

**Issue:** Migration uses `postgresql.JSONB(astext_type=sa.Text())` but the plan (Step 1) says "JSON, default=dict". Rule 129 (migrations must match model intent) is at risk of a mismatch if the plan is re-read without knowing JSONB was chosen.

**Suggestion:** No change needed to the migration itself (JSONB is correct for enrichment payloads), but the plan should be updated to reflect this.

File: src/db/models/customer_enrichment.py [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    """Third-party customer data enrichment record."""
+
+    __tablename__ = "customer_enrichments"
+
+    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
+    customer_id: Mapped[int] = mapped_column(
+        Integer, ForeignKey("customers.id", ondelete="CASCADE"), nullable=False, index=True
+    )
+    provider: Mapped[str] = mapped_column(String(100), nullable=False)
+    raw_data_json: Mapped[dict] = mapped_column(JSONB, default=dict, nullable=False)
```
Comment (yihua-zhuo) [inline src/db/models/customer_enrichment.py:22]: **WARNING** (suggestion)

**Issue:** The `raw_data_json` column declares `default=dict` with `nullable=False`, but SQLAlchemy's Python-side default only fires at INSERT time, not on `__init__` construction. This means `CustomerEnrichmentModel(customer_id=1, provider='test')` creates an object where `raw_data_json` is not set to `{}` until the session flushes. The `to_dict()` guards with `or {}`, but the type annotation `Mapped[dict]` is inaccurate if `raw_data_json` can be absent at the Python level.

**Suggestion:** Either change the type to `Mapped[dict | None]` and keep `or {}` in `to_dict()`, or remove `default=dict` and document that `raw_data_json` must always be explicitly set by callers. The pattern in `AutomationRuleModel` (which uses the same `default=list`/`Mapped[list]` combo) has the same latent issue.

File: src/db/models/customer_enrichment.py [hunk 9]
```diff
@@ -0,0 +1,44 @@
+"""Customer enrichment ORM model."""
+
+from datetime import datetime
```
Comment (yihua-zhuo) [inline src/db/models/customer_enrichment.py:3]: **INFO** (suggestion)

**Issue:** The `datetime` import on line 3 is unused — `datetime` is referenced in annotations as a class-level string ('forward reference') but never imported for runtime use.

**Suggestion:** Remove `from datetime import datetime` since it is not needed at runtime (all `datetime` references are in class-level annotations).

File: tests/unit/test_customer_enrichment_model.py [hunk 10]
```diff
@@ -0,0 +1,117 @@
+"""Unit tests for CustomerEnrichment ORM model."""
+
+from db.base import Base
+
+# Prefer the already-registered class when db.models.__init__.py loaded it
```
Comment (yihua-zhuo) [inline tests/unit/test_customer_enrichment_model.py:5]: **WARNING** (suggestion)

**Issue:** Lines 5-23 implement a fragile fallback import guard checking `Base.metadata.tables` before attempting filesystem loading. This pattern is not used by any other test in the project (e.g. `test_opportunity_model.py` uses a simple direct import). It adds unnecessary complexity and suggests the author was unsure whether the model would be auto-discovered.

**Suggestion:** Remove the conditional import block (lines 5-23) and use a simple direct import: `from db.models.customer_enrichment import CustomerEnrichmentModel`. The auto-discovery in `db/models/__init__.py` handles registration via `pkgutil.iter_modules`.

**Comment by yihua-zhuo** [comment]
## Claude Code Review (inline)

**Status:** `fail`
**Inline comments posted:** 10

**Summary:** Reviewed 4 file(s) between origin/master and origin/auto/issue-510.

### Issues without a usable file/line
- `alembic/versions/f18b406b982a_create_customer_enrichments.py`: Migration has no `tenant_id` column. Rule 126 (tenant isolation mandatory on every data path) and Rule 138 (tenant-owned tables need a real FK to `tenants.id`) are violated. Without a `tenant_id`, cross-tenant queries via the `customer_id` FK cannot be filtered at the SQL level.
- `src/db/models/customer_enrichment.py`: Model has no `tenant_id` column. Every other tenant-owned table in this codebase (`automation_rules`, `automation_logs`, `opportunities`, `activities`, etc.) has an explicit `tenant_id` column. Rule 126 makes tenant isolation mandatory on every data path, and Rule 138 requires a proper FK to `tenants.id` for tenant-owned ORM models.
- `tests/unit/test_customer_enrichment_model.py`: Tests do not exercise tenant isolation at all. Rule 134 requires unit mocks for tenant-owned data to enforce `tenant_id` on reads/writes/aggregates, and Rule 126 requires cross-tenant negative tests for any new multi-tenant data path. `CustomerEnrichmentModel` lacks a `tenant_id` column (see model review), so there is no isolation surface to test — this compounds the model's missing `tenant_id` issue.
- `tests/unit/test_customer_enrichment_model.py`: Tests instantiate ORM objects without a session and do not assert `created_at`/`updated_at` values, even though these fields are `server_default=func.now()` (populated by the DB on INSERT) and the Python object would have `None` for them before flush. `test_create_enrichment_defaults` verifies all fields but silently skips `created_at` and `updated_at` assertions.
